SPEC.md

```diff
@@ -77,33 +78,58 @@ Network configuration in JSON format is streamed to the plugin through stdin. Th
 
 ### Result
 
-Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. This should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details).
+Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. The `ip` and `dns` items should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details) except that the plugin should fill in the `interface` indexes appropriately, which are missing from IPAM plugin output since IPAM plugins should be unaware of interfaces.
 
 ```
 {
   "cniVersion": "0.1.0",
-  "ip4": {
-    "ip": <ipv4-and-subnet-in-CIDR>,
-    "gateway": <ipv4-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv4-routes>    (optional)
-  },
-  "ip6": {
-    "ip": <ipv6-and-subnet-in-CIDR>,
-    "gateway": <ipv6-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv6-routes>    (optional)
-  },
+  "interfaces": [
```

Comment:
> But the part that I'm missing is how will you set up traffic shaping if the plugin can return whatever it wants in the interfaces section. Will you try to introspect the "host" interfaces to see what they are and where they might be connected to?

Yes, that's what I would do in Kubernetes for now, for the simple cases.  If one of the returned host interfaces is a bridge, and the plugin has requested shaping, then that's the simplest way.  But this may actually be best handled by a CNI plugin, since then you don't need hints back to Kubernetes that shaping should be enabled.

> rkt implements port mapping via DNAT to container-ip, assuming that it's reachable from the host.

Good to know, but is that still true if rkt isn't setting up the container network, instead if that is being set up from Kubernetes via CNI plugins before the container is run with rkt?  Or is the DNAT stuff completely separate from any of the rkt network code?

> I know very little about traffic shaping but I would think you'd want to do that to the interface inside the container. For example, I doubt you can set tc on a slave of macvlan from the host side.

Current shaping using TC in Kubernetes uses the main host bridge interface (to set up the qdisc on the parent interface).  Shaping with OVS uses the host-side veth.